plat/common/include/x86/apic.h

```diff
@@ -44,30 +44,43 @@
 
 #include <errno.h>
 
-static inline int x2apic_enable(void)
+static inline int apic_enable(void)
 {
-	__u32 eax, ebx, ecx, edx;
+	__u32 eax, ebx __maybe_unused, ecx __maybe_unused, edx;
 
-	/* Check for x2APIC support */
-	cpuid(1, 0, &eax, &ebx, &ecx, &edx);
-	if (!(ecx & X86_CPUID1_ECX_x2APIC))
-		return -ENOTSUP;
-
-	/* Check if APIC is active */
+	/*
+	 * We expect to be started in xAPIC mode. This means firmware has not
+	 * irreversibly disabled the APIC and not yet enabled x2APIC mode.
+	 */
 	rdmsr(APIC_MSR_BASE, &eax, &edx);
-	if (!(eax & APIC_BASE_EN))
-		return -ENOTSUP;
+	UK_BUGON(!(eax & APIC_BASE_EN) || (eax & APIC_BASE_EXTD));
```

Comment:
this check should only be done if the CPU has x2apic support

Reply:
I would assume `APIC_BASE_EXTD` to be set to zero in case x2apic mode is not supported (at least that is what qemu seems to do), but it is probably better to not rely on it.